File: app/src/main/java/com/demos/kotlin/activity/charts/BarChartActivity.java
package com.demos.kotlin.activity.charts;

import android.graphics.RectF;
import android.os.Bundle;
import android.view.WindowManager;

import com.demos.kotlin.R;
import com.demos.kotlin.activity.DemoBase;
import com.demos.kotlin.bean.MarkerDataBean;
import com.demos.kotlin.views.charts.CombinedMarkerView;
import com.demos.kotlin.views.charts.DayAxisValueFormatter;
import com.demos.kotlin.views.charts.MyValueFormatter;
import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.Legend.LegendForm;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.XAxis.XAxisPosition;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.components.YAxis.AxisDependency;
import com.github.mikephil.charting.components.YAxis.YAxisLabelPosition;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.formatter.ValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.listener.OnChartValueSelectedListener;
import com.github.mikephil.charting.utils.MPPointF;

import java.util.ArrayList;
import java.util.List;

/**
 * 柱状图
 */
public class BarChartActivity extends DemoBase implements OnChartValueSelectedListener {

    private BarChart chart;
    private CombinedMarkerView mv;
    private List<MarkerDataBean> markerDataBeans;
    private BarDataSet set1;
    private ArrayList<BarEntry> values;//柱状图，填充横竖坐标值，两个属性

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_bar_chart);

        setTitle("BarChartActivity");

        chart = findViewById(R.id.chart1);
        chart.setOnChartValueSelectedListener(this);

        chart.setDrawBarShadow(false);
        chart.setDrawValueAboveBar(true);

        chart.getDescription().setEnabled(false);

        // if more than 60 entries are displayed in the chart, no values will be
        // drawn
        chart.setMaxVisibleValueCount(60);

        // scaling can now only be done on x- and y-axis separately
        chart.setPinchZoom(false);

        chart.setDrawGridBackground(false);
        // chart.setDrawYLabels(false);

        ValueFormatter xAxisFormatter = new DayAxisValueFormatter(chart);

        XAxis xAxis = chart.getXAxis();
        xAxis.setPosition(XAxisPosition.BOTTOM);
        xAxis.setDrawGridLines(false);
        xAxis.setGranularity(1f); // only intervals of 1 day
        xAxis.setLabelCount(7);
        xAxis.setValueFormatter(xAxisFormatter);

        ValueFormatter custom = new MyValueFormatter("$");

        YAxis leftAxis = chart.getAxisLeft();
        leftAxis.setLabelCount(8, false);
        leftAxis.setValueFormatter(custom);
        leftAxis.setPosition(YAxisLabelPosition.OUTSIDE_CHART);
        leftAxis.setSpaceTop(15f);
        leftAxis.setAxisMinimum(0f); // this replaces setStartAtZero(true)

        YAxis rightAxis = chart.getAxisRight();
        rightAxis.setDrawGridLines(false);
        rightAxis.setLabelCount(8, false);
        rightAxis.setValueFormatter(custom);
        rightAxis.setSpaceTop(15f);
        rightAxis.setAxisMinimum(0f); // this replaces setStartAtZero(true)

        Legend l = chart.getLegend();
        l.setVerticalAlignment(Legend.LegendVerticalAlignment.BOTTOM);
        l.setHorizontalAlignment(Legend.LegendHorizontalAlignment.LEFT);
        l.setOrientation(Legend.LegendOrientation.HORIZONTAL);
        l.setDrawInside(false);
        l.setForm(LegendForm.SQUARE);
        l.setFormSize(9f);
        l.setTextSize(11f);
        l.setXEntrySpace(4f);
        setData();

        chart.getLegend().setEnabled(true);
    }

    //TODO 网络请求之后，设置图表，或者是刷新数据
    private void setData() {
        //填数据
        float start = 1f;
        if (values == null) {
            values = new ArrayList<>();
        } else {
            values.clear();
        }
        for (int i = (int) start; i < start + 10; i++) {
            float val = (float) (Math.random() * (20 + 1));
            values.add(new BarEntry(i, val));
        }
        if (chart.getData() != null && chart.getData().getDataSetCount() > 0) {
            set1 = (BarDataSet) chart.getData().getDataSetByIndex(0);
            set1.setValues(values);
            chart.getData().notifyDataChanged();
            chart.notifyDataSetChanged();
        } else {
            set1 = new BarDataSet(values, "The year 2017");
            set1.setDrawIcons(false);
            ArrayList<IBarDataSet> dataSets = new ArrayList<>();
            dataSets.add(set1);
            final BarData data = new BarData(dataSets);
//            data.setValueTextSize(10f);
            data.setBarWidth(0.4f);
            chart.setData(data);
            mv = new CombinedMarkerView(this);
            markerDataBeans = new ArrayList<>();//
            mv.setChartView(chart); // For bounds control
            chart.setMarker(mv); // Set the marker to the chart
        }
    }


    @Override
    protected void saveToGallery() {
        saveToGallery(chart, "BarChartActivity");
    }

    @Override
    public void onValueSelected(Entry e, Highlight h) {

        if (e == null)
            return;
        //刷新markerView的数据
        markerDataBeans.clear();
        markerDataBeans.add(new MarkerDataBean(set1.getLabel(), String.valueOf(values.get((int) (h.getX() - 1)).getY())));
        mv.setData(markerDataBeans, "marker的title", values.size());
        chart.getBarBounds((BarEntry) e, new RectF());
        MPPointF position = chart.getPosition(e, AxisDependency.LEFT);
        mv.refreshContent(e, h);
        MPPointF.recycleInstance(position);
    }

    @Override
    public void onNothingSelected() {
    }
}

File: app/src/main/java/com/demos/kotlin/activity/kotlinchart/BarChartActivity.kt
package com.demos.kotlin.activity.kotlinchart

import android.graphics.RectF
import android.os.Bundle
import android.view.WindowManager
import com.demos.kotlin.R
import com.demos.kotlin.activity.DemoBase
import com.demos.kotlin.bean.MarkerDataBean
import com.demos.kotlin.views.charts.CombinedMarkerView
import com.demos.kotlin.views.charts.DayAxisValueFormatter
import com.demos.kotlin.views.charts.MyValueFormatter
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.Legend.LegendForm
import com.github.mikephil.charting.components.XAxis.XAxisPosition
import com.github.mikephil.charting.components.YAxis.AxisDependency
import com.github.mikephil.charting.components.YAxis.YAxisLabelPosition
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.github.mikephil.charting.listener.OnChartValueSelectedListener
import com.github.mikephil.charting.utils.MPPointF
import java.util.*

/**
 * 柱状图
 */
class BarChartActivity : DemoBase(), OnChartValueSelectedListener {
    private lateinit var chart: BarChart
    private lateinit var mv: CombinedMarkerView
    private var markerDataBeans: MutableList<MarkerDataBean>? = null
    private var set1: BarDataSet? = null
    private var values //柱状图，填充横竖坐标值，两个属性
            : ArrayList<BarEntry>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN)
        setContentView(R.layout.activity_bar_chart)
        title = "BarChartActivity"
        chart = findViewById(R.id.chart1)
        chart.setOnChartValueSelectedListener(this)
        chart.setDrawBarShadow(false)
        chart.setDrawValueAboveBar(true)
        chart.getDescription().isEnabled = false

        // if more than 60 entries are displayed in the chart, no values will be
        // drawn
        chart.setMaxVisibleValueCount(60)

        // scaling can now only be done on x- and y-axis separately
        chart.setPinchZoom(false)
        chart.setDrawGridBackground(false)
        // chart.setDrawYLabels(false);
        val xAxisFormatter: ValueFormatter = DayAxisValueFormatter(chart)
        val xAxis = chart.getXAxis()
        xAxis.position = XAxisPosition.BOTTOM
        xAxis.setDrawGridLines(false)
        xAxis.granularity = 1f // only intervals of 1 day
        xAxis.labelCount = 7
        xAxis.valueFormatter = xAxisFormatter
        val custom: ValueFormatter = MyValueFormatter("$")
        val leftAxis = chart.getAxisLeft()
        leftAxis.setLabelCount(8, false)
        leftAxis.valueFormatter = custom
        leftAxis.setPosition(YAxisLabelPosition.OUTSIDE_CHART)
        leftAxis.spaceTop = 15f
        leftAxis.axisMinimum = 0f // this replaces setStartAtZero(true)
        val rightAxis = chart.getAxisRight()
        rightAxis.setDrawGridLines(false)
        rightAxis.setLabelCount(8, false)
        rightAxis.valueFormatter = custom
        rightAxis.spaceTop = 15f
        rightAxis.axisMinimum = 0f // this replaces setStartAtZero(true)
        val l = chart.getLegend()
        l.verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
        l.horizontalAlignment = Legend.LegendHorizontalAlignment.LEFT
        l.orientation = Legend.LegendOrientation.HORIZONTAL
        l.setDrawInside(false)
        l.form = LegendForm.SQUARE
        l.formSize = 9f
        l.textSize = 11f
        l.xEntrySpace = 4f
        setData()
        chart.getLegend().isEnabled = true
    }

    //TODO 网络请求之后，设置图表，或者是刷新数据
    private fun setData() {
        //填数据
        val start = 1f
        if (values == null) {
            values = ArrayList()
        } else {
            values!!.clear()
        }
        var i = start.toInt()
        while (i < start + 10) {
            val value = (Math.random() * (20 + 1)).toFloat()
            values!!.add(BarEntry(i.toFloat(), value))
            i++
        }
        if (chart!!.data != null && chart!!.data.dataSetCount > 0) {
            set1 = chart!!.data.getDataSetByIndex(0) as BarDataSet
            set1!!.values = values
            chart!!.data.notifyDataChanged()
            chart!!.notifyDataSetChanged()
        } else {
            set1 = BarDataSet(values, "The year 2017")
            set1!!.setDrawIcons(false)
            val dataSets = ArrayList<IBarDataSet>()
            dataSets.add(set1!!)
            val data = BarData(dataSets)
            //            data.setValueTextSize(10f);
            data.barWidth = 0.4f
            chart!!.data = data
            mv = CombinedMarkerView(this)
            markerDataBeans = ArrayList() //
            mv!!.chartView = chart // For bounds control
            chart!!.marker = mv // Set the marker to the chart
        }
    }

    override fun saveToGallery() {
        saveToGallery(chart, "BarChartActivity")
    }

    override fun onValueSelected(e: Entry, h: Highlight) {
        if (e == null) return
        //刷新markerView的数据
        markerDataBeans!!.clear()
        markerDataBeans!!.add(MarkerDataBean(set1!!.label, values!![(h.x - 1).toInt()].y.toString()))
        mv!!.setData(markerDataBeans, "marker的title", values!!.size)
        chart!!.getBarBounds(e as BarEntry, RectF())
        val position = chart!!.getPosition(e, AxisDependency.LEFT)
        mv!!.refreshContent(e, h)
        MPPointF.recycleInstance(position)
    }

    override fun onNothingSelected() {}
}